### ISSUE TITLE
Variance unexpected behavior

### ISSUE THREAD

**Issue by OriRoth** [microsoft/pyright#4210]
I get unexpected behavior when checking programs that use (contra)variance.

For example:
```python
from typing import TypeVar, Generic
Z = TypeVar("Z", contravariant=True)
class N(Generic[Z]): ...
class C(N[N["C"]]): ...
_: N[C] = C()
```
This program should not type check. Mypy and Pyre crash as expected. Pyright reports no error (false negative).

---
```python
from typing import TypeVar, Generic, Any
Z = TypeVar("Z", contravariant=True)
class N(Generic[Z]): ...
X = TypeVar("X")
class C(Generic[X], N[N["C[C[X]]"]]): ...
_: N[C[Any]] = C[Any]()
```
This is a similar case. Pyright returns an error, which is fine, but the error message is probably wrong or incomplete:
```
error: Expression of type "C[Any]" cannot be assigned to declared type "N[C[Any]]"
    "C[Any]" is incompatible with "N[C[Any]]"
      TypeVar "Z@N" is contravariant
        "C[Any]" is incompatible with "N[C[C[Any]]]"
          TypeVar "Z@N" is contravariant
            "C[C[Any]]" is incompatible with "N[C[C[Any]]]" (reportGeneralTypeIssues)
```
---
This is a more complicated example. This program should type check. Mypy and Pyre report no error as expected. Pyright reports an error (false positive).
```python
from typing import TypeVar, Generic
T = TypeVar("T", contravariant=True)
class Z: ...
class N(Generic[T]): ...
class ML(Generic[T]): ...
class MR(Generic[T]): ...
class L_a(Generic[T]): ...
class L_c(Generic[T]): ...
class L_b(Generic[T]): ...
class L___TAPE_END__(Generic[T]): ...
class QLR_q0(Generic[T]): ...
class QLR_q1(Generic[T]): ...
class QLR_q2(Generic[T]): ...
class QLR_q8(Generic[T]): ...
class QLR_q3(Generic[T]): ...
class QLR_q4(Generic[T]): ...
class QLR_q5(Generic[T]): ...
class QLR_q6(Generic[T]): ...
class QLR_q7(Generic[T]): ...
class QRL_q0(Generic[T]): ...
class QRL_q1(Generic[T]): ...
class QRL_q2(Generic[T]): ...
class QRL_q8(Generic[T]): ...
class QRL_q3(Generic[T]): ...
class QRL_q4(Generic[T]): ...
class QRL_q5(Generic[T]): ...
class QRL_q6(Generic[T]): ...
class QRL_q7(Generic[T]): ...
class E(Generic[T], QLR_q0["N[QRW_q0[E[E[T]]]]"], QLR_q1["N[QRW_q1[E[E[T]]]]"], QLR_q2["N[QRW_q2[E[E[T]]]]"], QLR_q8["N[QRW_q8[E[E[T]]]]"], QLR_q3["N[QRW_q3[E[E[T]]]]"], QLR_q4["N[QRW_q4[E[E[T]]]]"], QLR_q5["N[QRW_q5[E[E[T]]]]"], QLR_q6["N[QRW_q6[E[E[T]]]]"], QLR_q7["N[QRW_q7[E[E[T]]]]"], QRL_q0["N[QLW_q0[E[E[T]]]]"], QRL_q1["N[QLW_q1[E[E[T]]]]"], QRL_q2["N[QLW_q2[E[E[T]]]]"], QRL_q8["N[QLW_q8[E[E[T]]]]"], QRL_q3["N[QLW_q3[E[E[T]]]]"], QRL_q4["N[QLW_q4[E[E[T]]]]"], QRL_q5["N[QLW_q5[E[E[T]]]]"], QRL_q6["N[QLW_q6[E[E[T]]]]"], QRL_q7["N[QLW_q7[E[E[T]]]]"]): ...
class QLW_q0(Generic[T], ML["N[QL_q0[T]]"], MR["N[QLW_q0[MR[N[T]]]]"], L_a["N[QLW_q0[L_a[N[T]]]]"], L_c["N[QLW_q0[L_c[N[T]]]]"], L_b["N[QLW_q0[L_b[N[T]]]]"], L___TAPE_END__["N[QLW_q0[L___TAPE_END__[N[T]]]]"], E["QLR_q0[N[T]]"]): ...
class QRW_q0(Generic[T], MR["N[QR_q0[T]]"], ML["N[QRW_q0[ML[N[T]]]]"], L_a["N[QRW_q0[L_a[N[T]]]]"], L_c["N[QRW_q0[L_c[N[T]]]]"], L_b["N[QRW_q0[L_b[N[T]]]]"], L___TAPE_END__["N[QRW_q0[L___TAPE_END__[N[T]]]]"], E["QRL_q0[N[T]]"]): ...
class QLW_q1(Generic[T], ML["N[QL_q1[T]]"], MR["N[QLW_q1[MR[N[T]]]]"], L_a["N[QLW_q1[L_a[N[T]]]]"], L_c["N[QLW_q1[L_c[N[T]]]]"], L_b["N[QLW_q1[L_b[N[T]]]]"], L___TAPE_END__["N[QLW_q1[L___TAPE_END__[N[T]]]]"], E["QLR_q1[N[T]]"]): ...
class QRW_q1(Generic[T], MR["N[QR_q1[T]]"], ML["N[QRW_q1[ML[N[T]]]]"], L_a["N[QRW_q1[L_a[N[T]]]]"], L_c["N[QRW_q1[L_c[N[T]]]]"], L_b["N[QRW_q1[L_b[N[T]]]]"], L___TAPE_END__["N[QRW_q1[L___TAPE_END__[N[T]]]]"], E["QRL_q1[N[T]]"]): ...
class QLW_q2(Generic[T], ML["N[QL_q2[T]]"], MR["N[QLW_q2[MR[N[T]]]]"], L_a["N[QLW_q2[L_a[N[T]]]]"], L_c["N[QLW_q2[L_c[N[T]]]]"], L_b["N[QLW_q2[L_b[N[T]]]]"], L___TAPE_END__["N[QLW_q2[L___TAPE_END__[N[T]]]]"], E["QLR_q2[N[T]]"]): ...
class QRW_q2(Generic[T], MR["N[QR_q2[T]]"], ML["N[QRW_q2[ML[N[T]]]]"], L_a["N[QRW_q2[L_a[N[T]]]]"], L_c["N[QRW_q2[L_c[N[T]]]]"], L_b["N[QRW_q2[L_b[N[T]]]]"], L___TAPE_END__["N[QRW_q2[L___TAPE_END__[N[T]]]]"], E["QRL_q2[N[T]]"]): ...
class QLW_q8(Generic[T], ML["N[QL_q8[T]]"], MR["N[QLW_q8[MR[N[T]]]]"], L_a["N[QLW_q8[L_a[N[T]]]]"], L_c["N[QLW_q8[L_c[N[T]]]]"], L_b["N[QLW_q8[L_b[N[T]]]]"], L___TAPE_END__["N[QLW_q8[L___TAPE_END__[N[T]]]]"], E["E[Z]"]): ...
class QRW_q8(Generic[T], MR["N[QR_q8[T]]"], ML["N[QRW_q8[ML[N[T]]]]"], L_a["N[QRW_q8[L_a[N[T]]]]"], L_c["N[QRW_q8[L_c[N[T]]]]"], L_b["N[QRW_q8[L_b[N[T]]]]"], L___TAPE_END__["N[QRW_q8[L___TAPE_END__[N[T]]]]"], E["E[Z]"]): ...
class QLW_q3(Generic[T], ML["N[QL_q3[T]]"], MR["N[QLW_q3[MR[N[T]]]]"], L_a["N[QLW_q3[L_a[N[T]]]]"], L_c["N[QLW_q3[L_c[N[T]]]]"], L_b["N[QLW_q3[L_b[N[T]]]]"], L___TAPE_END__["N[QLW_q3[L___TAPE_END__[N[T]]]]"], E["QLR_q3[N[T]]"]): ...
class QRW_q3(Generic[T], MR["N[QR_q3[T]]"], ML["N[QRW_q3[ML[N[T]]]]"], L_a["N[QRW_q3[L_a[N[T]]]]"], L_c["N[QRW_q3[L_c[N[T]]]]"], L_b["N[QRW_q3[L_b[N[T]]]]"], L___TAPE_END__["N[QRW_q3[L___TAPE_END__[N[T]]]]"], E["QRL_q3[N[T]]"]): ...
class QLW_q4(Generic[T], ML["N[QL_q4[T]]"], MR["N[QLW_q4[MR[N[T]]]]"], L_a["N[QLW_q4[L_a[N[T]]]]"], L_c["N[QLW_q4[L_c[N[T]]]]"], L_b["N[QLW_q4[L_b[N[T]]]]"], L___TAPE_END__["N[QLW_q4[L___TAPE_END__[N[T]]]]"], E["QLR_q4[N[T]]"]): ...
class QRW_q4(Generic[T], MR["N[QR_q4[T]]"], ML["N[QRW_q4[ML[N[T]]]]"], L_a["N[QRW_q4[L_a[N[T]]]]"], L_c["N[QRW_q4[L_c[N[T]]]]"], L_b["N[QRW_q4[L_b[N[T]]]]"], L___TAPE_END__["N[QRW_q4[L___TAPE_END__[N[T]]]]"], E["QRL_q4[N[T]]"]): ...
class QLW_q5(Generic[T], ML["N[QL_q5[T]]"], MR["N[QLW_q5[MR[N[T]]]]"], L_a["N[QLW_q5[L_a[N[T]]]]"], L_c["N[QLW_q5[L_c[N[T]]]]"], L_b["N[QLW_q5[L_b[N[T]]]]"], L___TAPE_END__["N[QLW_q5[L___TAPE_END__[N[T]]]]"], E["QLR_q5[N[T]]"]): ...
class QRW_q5(Generic[T], MR["N[QR_q5[T]]"], ML["N[QRW_q5[ML[N[T]]]]"], L_a["N[QRW_q5[L_a[N[T]]]]"], L_c["N[QRW_q5[L_c[N[T]]]]"], L_b["N[QRW_q5[L_b[N[T]]]]"], L___TAPE_END__["N[QRW_q5[L___TAPE_END__[N[T]]]]"], E["QRL_q5[N[T]]"]): ...
class QLW_q6(Generic[T], ML["N[QL_q6[T]]"], MR["N[QLW_q6[MR[N[T]]]]"], L_a["N[QLW_q6[L_a[N[T]]]]"], L_c["N[QLW_q6[L_c[N[T]]]]"], L_b["N[QLW_q6[L_b[N[T]]]]"], L___TAPE_END__["N[QLW_q6[L___TAPE_END__[N[T]]]]"], E["QLR_q6[N[T]]"]): ...
class QRW_q6(Generic[T], MR["N[QR_q6[T]]"], ML["N[QRW_q6[ML[N[T]]]]"], L_a["N[QRW_q6[L_a[N[T]]]]"], L_c["N[QRW_q6[L_c[N[T]]]]"], L_b["N[QRW_q6[L_b[N[T]]]]"], L___TAPE_END__["N[QRW_q6[L___TAPE_END__[N[T]]]]"], E["QRL_q6[N[T]]"]): ...
class QLW_q7(Generic[T], ML["N[QL_q7[T]]"], MR["N[QLW_q7[MR[N[T]]]]"], L_a["N[QLW_q7[L_a[N[T]]]]"], L_c["N[QLW_q7[L_c[N[T]]]]"], L_b["N[QLW_q7[L_b[N[T]]]]"], L___TAPE_END__["N[QLW_q7[L___TAPE_END__[N[T]]]]"], E["QLR_q7[N[T]]"]): ...
class QRW_q7(Generic[T], MR["N[QR_q7[T]]"], ML["N[QRW_q7[ML[N[T]]]]"], L_a["N[QRW_q7[L_a[N[T]]]]"], L_c["N[QRW_q7[L_c[N[T]]]]"], L_b["N[QRW_q7[L_b[N[T]]]]"], L___TAPE_END__["N[QRW_q7[L___TAPE_END__[N[T]]]]"], E["QRL_q7[N[T]]"]): ...
class QL_q0(Generic[T], L_a["N[QLW_q1[L_c[N[MR[N[T]]]]]]"], L_b["N[QLW_q5[L_c[N[MR[N[T]]]]]]"], L___TAPE_END__["N[QLW_q8[L___TAPE_END__[N[L_c[N[MR[N[T]]]]]]]]"], L_c["N[QLW_q8[L_c[N[MR[N[T]]]]]]"]): ...
class QR_q0(Generic[T], L_a["N[QRW_q1[MR[N[L_c[N[T]]]]]]"], L_b["N[QRW_q5[MR[N[L_c[N[T]]]]]]"], L___TAPE_END__["N[QRW_q8[L___TAPE_END__[N[MR[N[L_c[N[T]]]]]]]]"], L_c["N[QRW_q8[MR[N[L_c[N[T]]]]]]"]): ...
class QL_q1(Generic[T], L_a["N[QLW_q2[L_a[N[MR[N[T]]]]]]"], L_b["N[QLW_q2[L_b[N[MR[N[T]]]]]]"], L___TAPE_END__["N[QLW_q8[L___TAPE_END__[N[L_c[N[MR[N[T]]]]]]]]"], L_c["N[QLW_q8[L_c[N[MR[N[T]]]]]]"]): ...
class QR_q1(Generic[T], L_a["N[QRW_q2[MR[N[L_a[N[T]]]]]]"], L_b["N[QRW_q2[MR[N[L_b[N[T]]]]]]"], L___TAPE_END__["N[QRW_q8[L___TAPE_END__[N[MR[N[L_c[N[T]]]]]]]]"], L_c["N[QRW_q8[MR[N[L_c[N[T]]]]]]"]): ...
class QL_q2(Generic[T], L_a["N[QLW_q2[L_a[N[MR[N[T]]]]]]"], L_b["N[QLW_q2[L_b[N[MR[N[T]]]]]]"], L___TAPE_END__["N[QLW_q3[L___TAPE_END__[N[ML[N[L_c[N[T]]]]]]]]"], L_c["N[QLW_q3[ML[N[L_c[N[T]]]]]]"]): ...
class QR_q2(Generic[T], L_a["N[QRW_q2[MR[N[L_a[N[T]]]]]]"], L_b["N[QRW_q2[MR[N[L_b[N[T]]]]]]"], L___TAPE_END__["N[QRW_q3[L___TAPE_END__[N[L_c[N[ML[N[T]]]]]]]]"], L_c["N[QRW_q3[L_c[N[ML[N[T]]]]]]"]): ...
class QL_q8(Generic[T]): ...
class QR_q8(Generic[T]): ...
class QL_q3(Generic[T], L_a["N[QLW_q4[ML[N[L_c[N[T]]]]]]"]): ...
class QR_q3(Generic[T], L_a["N[QRW_q4[L_c[N[ML[N[T]]]]]]"]): ...
class QL_q4(Generic[T], L_a["N[QLW_q4[ML[N[L_a[N[T]]]]]]"], L_b["N[QLW_q4[ML[N[L_b[N[T]]]]]]"], L___TAPE_END__["N[QLW_q0[L___TAPE_END__[N[L_c[N[MR[N[T]]]]]]]]"], L_c["N[QLW_q0[L_c[N[MR[N[T]]]]]]"]): ...
class QR_q4(Generic[T], L_a["N[QRW_q4[L_a[N[ML[N[T]]]]]]"], L_b["N[QRW_q4[L_b[N[ML[N[T]]]]]]"], L___TAPE_END__["N[QRW_q0[L___TAPE_END__[N[MR[N[L_c[N[T]]]]]]]]"], L_c["N[QRW_q0[MR[N[L_c[N[T]]]]]]"]): ...
class QL_q5(Generic[T], L_a["N[QLW_q6[L_a[N[MR[N[T]]]]]]"], L_b["N[QLW_q6[L_b[N[MR[N[T]]]]]]"], L___TAPE_END__["N[QLW_q8[L___TAPE_END__[N[L_c[N[MR[N[T]]]]]]]]"], L_c["N[QLW_q8[L_c[N[MR[N[T]]]]]]"]): ...
class QR_q5(Generic[T], L_a["N[QRW_q6[MR[N[L_a[N[T]]]]]]"], L_b["N[QRW_q6[MR[N[L_b[N[T]]]]]]"], L___TAPE_END__["N[QRW_q8[L___TAPE_END__[N[MR[N[L_c[N[T]]]]]]]]"], L_c["N[QRW_q8[MR[N[L_c[N[T]]]]]]"]): ...
class QL_q6(Generic[T], L_a["N[QLW_q6[L_a[N[MR[N[T]]]]]]"], L_b["N[QLW_q6[L_b[N[MR[N[T]]]]]]"], L___TAPE_END__["N[QLW_q7[L___TAPE_END__[N[ML[N[L_c[N[T]]]]]]]]"], L_c["N[QLW_q7[ML[N[L_c[N[T]]]]]]"]): ...
class QR_q6(Generic[T], L_a["N[QRW_q6[MR[N[L_a[N[T]]]]]]"], L_b["N[QRW_q6[MR[N[L_b[N[T]]]]]]"], L___TAPE_END__["N[QRW_q7[L___TAPE_END__[N[L_c[N[ML[N[T]]]]]]]]"], L_c["N[QRW_q7[L_c[N[ML[N[T]]]]]]"]): ...
class QL_q7(Generic[T], L_b["N[QLW_q4[ML[N[L_c[N[T]]]]]]"]): ...
class QR_q7(Generic[T], L_b["N[QRW_q4[L_c[N[ML[N[T]]]]]]"]): ...
_: E[E[Z]] = QRW_q0[L___TAPE_END__[N[L_a[N[L_b[N[L_b[N[L_a[N[L_b[N[L_b[N[L_a[N[MR[N[L___TAPE_END__[N[E[E[Z]]]]]]]]]]]]]]]]]]]]]]]()
```
You can use [my project](https://github.com/OriRoth/python-typing-machines) to generate more examples such as this.

**Comment by erictraut** [comment]
> Mypy and Pyre crash as expected

Seriously? You expect a type checker to crash in this situation? I confirmed that mypy does crash (Segmentation fault), so that's a bug in mypy. If pyre also crashes, then I'd consider that a bug as well.

I think pyright is correct in its analysis, so I don't consider this a false negative.

Here are the steps a type checker must go through to evaluate type compatibility in this case:
1. Is `C` is a subtype of `N[C]`?
2. Since `C` derives from `N[N[C]]`, that's the same as asking whether `N[N[C]]` is a subtype of `N[C]`.
3. The type parameter for `N` is contravariant, so we need to determine whether the type argument `C` is subtype of type argument `N[C]`.
4. This is the same as asking whether `N[N[C]]` is a subtype of `N[C]`.

In step 4, we're asking the same question as in step 2, which means that we are recursing, and the answer is indeterminate. The other type checkers appear to not detect this recursion and simply crash (probably with a stack overflow). Pyright treats this indeterminate case as though it is allowed (which is what I would expect as a user).

My recommendation is that you avoid creating recursion like this in your types. It leads to indeterminate results (at best) and crashes (at worst).
